### ISSUE TITLE
How to fetch event codes of PERF_METRICS.FRONTEND_BOUND from icelakex_metrics.json file

### ISSUE THREAD

**Issue by Osmanyasal** [intel/perfmon#249]
Hello, 

TL;DR; I need event codes for PERF_METRICS.FRONTEND_BOUND, PERF_METRICS.BAD_SPECULATION etc. how to obtain them is there a script or a place to look at?

I would like to implement some portion of the TMAM analysis for educational purposes, although I have all the event codes for my architecture (ICL) I cannot find the corresponding events for FrontendBound although I've the formula regarding how it is calculated
From the file /ICL/metrics/icelakex_metrics.json I've the following description, Althoguh 1 of them has direct event code INT_MISC.UOP_DROPPING, others are unclear. 

I'd be glad if you help me,
Best.

Here's the json element

{
  "MetricName": "Frontend_Bound",
  "LegacyName": "metric_TMA_Frontend_Bound(%)",
  "Level": 1,
  "BriefDescription": "This category represents fraction of slots where the processor's Frontend undersupplies its Backend. Frontend denotes the first part of the processor core responsible to fetch operations that are executed later on by the Backend part. Within the Frontend; a branch predictor predicts the next address to fetch; cache-lines are fetched from the memory subsystem; parsed into instructions; and lastly decoded into micro-operations (uops). Ideally the Frontend can issue Pipeline_Width uops every cycle to the Backend. Frontend Bound denotes unutilized issue-slots when there is no Backend stall; i.e. bubbles where Frontend delivered no uops while Backend could have accepted them. For example; stalls due to instruction-cache misses would be categorized under Frontend Bound.",
  "UnitOfMeasure": "percent",
  "Events": [
    {
      "Name": "PERF_METRICS.FRONTEND_BOUND",
      "Alias": "a"
    },
    {
      "Name": "PERF_METRICS.BAD_SPECULATION",
      "Alias": "b"
    },
    {
      "Name": "PERF_METRICS.RETIRING",
      "Alias": "c"
    },
    {
      "Name": "PERF_METRICS.BACKEND_BOUND",
      "Alias": "d"
    },
    {
      "Name": "INT_MISC.UOP_DROPPING",
      "Alias": "e"
    },
    {
      "Name": "TOPDOWN.SLOTS:perf_metrics",
      "Alias": "f"
    }
  ],
  "Constants": [],
  "Formula": "100 * ( a / ( a + b + c + d ) - e / ( f ) )",
  "BaseFormula": " perf_metrics.frontend_bound / ( perf_metrics.frontend_bound + perf_metrics.bad_speculation + perf_metrics.retiring + perf_metrics.backend_bound ) - int_misc.uop_dropping / tma_info_thread_slots",
  "Category": "TMA",
  "CountDomain": "Slots",
  "Threshold": {
    "ThresholdMetrics": [
      {
        "Alias": "a",
        "Value": "metric_TMA_Frontend_Bound(%)"
      }
    ],
    "Formula": "a > 15",
    "BaseFormula": "metric_TMA_Frontend_Bound(%) > 15",
    "ThresholdIssues": ""
  },
  "ResolutionLevels": "THREAD, CORE, SOCKET, SYSTEM",
  "MetricGroup": "BvFB;BvIO;TmaL1;PGO",
  "LocateWith": " FRONTEND_RETIRED.LATENCY_GE_4"
}

**Comment by edwarddavidbaker** [comment]
Hi @Osmanyasal thanks for creating a ticket. Please refer to https://github.com/intel/perfmon/issues/225#issuecomment-2423192806 and https://github.com/intel/perfmon/issues/225#issuecomment-2423216781 . Feel free to re-open if you additional questions.

**Comment by Osmanyasal** [comment]
Hello @edwarddavidbaker, thank you for the response, I indeed have brief questions. 
I checked the original paper "A Top-Down method for performance analysis and counters architecture" by Ahmed Yasin.

--> In his implementations, there are many operations to obtain "front-end bound, be-bound" etc. however the link only monitors a few "direct" event codes like;
#define SLOTS               0x0400ull
#define RETIRING_SLOTS      0x8000ull
#define BAD_SPEC_SLOTS      0x8100ull
#define FE_BOUND_SLOTS      0x8200ull
#define BE_BOUND_SLOTS      0x8300ull

I also use perf_event_open and what he did is to create a group-leader with slots and group all these. when reading, it divides all other metrics to slots, which is different from what Ahmet suggests, how does this work?

// original implementation
![image](https://github.com/user-attachments/assets/6c3490d3-5bb0-455d-bcd6-dfe920a44d08)

 
--> I know pmu events are different per u-arhitecture, are these event codes (0x0400, 0x8000 etc) fix for each new-gen intel processor? I skimmed through the manual, I believe intel reserved some pre-defined event codes for TMA analysis.

Looking forward the answer
Best wishes,
Osman.

**Comment by captain5050** [comment]
Hi Osman,

If you are wanting to use Linux then perf has topdown analysis built in:
https://perfwiki.github.io/main/top-down-analysis/
You can also use `perf list -j` to dump out event encodings and metrics. If you get topdown data with something like `perf stat -M TopdownL1 <benchmark>` then there are verbose options (-v, -vv, -vvv) used like `perf stat -vv -M TopdownL1 <benchmark>` which will dump out the perf_event_open options used.

Thanks,
Ian

**Comment by Osmanyasal** [comment]
Hello @captain5050, thank you for the response. I am actually building a library for energy optimisation in mind but also one that supports performance analysis. 

Now it is time for me to implement TMA to increase use-cases. However I need to know whether the above events are definite among intel processors (which pmu events often are not) and whether the measurement is correct compared to the paper

Best
Osman.

**Comment by captain5050** [comment]
Thanks Osman. A few thoughts:

1. for C/C++ code perf's been trying to do this libperf separation, it should really have things like event parsing in it.
2. there's a python API for perf that has been traditionally used to process events for things like [flame graphs and the firefox profiler](https://perfwiki.github.io/main/tutorial/#flame-graph). [This patch series](https://lore.kernel.org/all/20241119011644.971342-1-irogers@google.com/) that just landed in the [development tree](https://git.kernel.org/pub/scm/linux/kernel/git/perf/perf-tools-next.git/) makes it so that you write python scripts doing perf like things using things like regular event names. I've not yet extended it to handle metrics, but there is already a bunch of [python metric code](https://git.kernel.org/pub/scm/linux/kernel/git/perf/perf-tools-next.git/tree/tools/perf/pmu-events/metric.py).
3. in the next linux release hwmon will be exposing new events for things like energy, power, .. [Here is the patch series](https://lore.kernel.org/lkml/20241109003759.473460-1-irogers@google.com/).
4. I'm aware of things like DRAM power consumption being something we should expose as perf events when the hardware supports it. There's work on other platforms to do things like per-core power analysis too.
